Baseline estimates
----------
----------

Collaborative Filtering data commonly displays significant biases. Some users consistently give higher ratings than others, and some items consistently receive higher ratings. To address these biases, it's standard practice to adjust the data by incorporating 'baseline estimates.'

The overall average rating is represented by $\mu$. A baseline estimate for an unknown rating ($r_{ui}$) is denoted by $b_{ui}$ and considers the biases of the specific user ($b_{u}$) and item ($b_{i}$):

Equation 1 $$b_{ui} = \mu + b_u + b_i$$
----------

The parameters $b_{u}$ and $b_{i}$ represent how much a particular user ('u') or item ('i') deviates from the average rating.

For instance, let's consider estimating the rating of the movie "Titanic" by a user named "Joe."

- $\mu$: the average rating across all movies is 3.7 stars.
- $b_{i}$: "Titanic" is generally rated 0.5 stars higher than the average.
- $b_{u}$: "Joe" tends to rate movies 0.3 stars lower than the average.

Therefore, the baseline estimate for "Joe's" rating of "Titanic" would be 3.9 stars: 

- $b_{ui}$ = 3.7 - 0.3 + 0.5 = 3.9

In order to estimate $b_{u}$ and $b_{i}$ one can solve the least squares problem:

 Equation 2 $$\min_{b} \sum_{(u,i) \in K} \left(r_{ui} - \mu - b_u - b_i\right)^2 + \lambda \left(\sum_u b_u^2 + \sum_i b_i^2\right)$$
---------------


Equation 2 is posed as an optimization problem, which can be divided into two parts.

The first term measures the squared error between the actual and predicted ratings:

$$\min_{b} \sum_{(u,i) \in K} \left(r_{ui} - \mu - b_u - b_i\right)^2$$

Here, we want to minimize the difference between the actual score ($r_{ui}$) and the estimate based on average effects ($-\mu - b_u - b_i$)

K is the set of pairs (u,i) where ratings are available.

This is the second term:

$$\lambda \left(\sum_u b_u^2 + \sum_i b_i^2\right)$$

This term penalizes large values of $b_u$ and $b_i$, which could overfit the training data (overfitting).

$\lambda$ is a regularization parameter that controls the weight of this penalty. A larger $\lambda$ means less flexibility, but less risk of overfitting.

Let's see how the optimization process works
--------------
-------------

To optimize the parameters $b_u$ and $b_i$, we take the partial derivatives of L with respect to each one.

The L function is the loss function that we want to minimize using optimization techniques such as gradient descent. By differentiating L with respect to parameters $b_u$ and $b_i$, we obtain the formulas necessary to update these parameters in a way that best fits the data.

### 1. The loss function
$ L $ is defined as:
$$
L = \sum_{(u, i) \in K} \left( r_{ui} - \mu - b_u - b_i \right)^2 + \lambda \left( \sum_u b_u^2 + \sum_i b_i^2 \right).
$$

We have already explained the different parts of $L$

### 2. Partial derivative of $ L $ with respect to $ b_u$
To calculate the partial derivative of $L$ with respect to $b_u$, we take only the terms that depend on $b_u$:
$$
\frac{\partial L}{\partial b_u} = \frac{\partial}{\partial b_u} \sum_{(u, i) \in K} \left( r_{ui} - \mu - b_u - b_i \right)^2 + \frac{\partial}{\partial b_u} \lambda \sum_u b_u^2.
$$

#### Quadratic error term:
#### Step 1: Apply the chain rule
The first term is a square, so we apply the chain rule. The derivative of a quadratic function $ f(x) = g(x)^2 $ is:
$$
\frac{\partial}{\partial x} g(x)^2 = 2 \cdot g(x) \cdot \frac{\partial g(x)}{\partial x}.
$$

In this case, we define:
$$
g(x) = r_{ui} - \mu - b_u - b_i.
$$

Therefore, when deriving:
$$
\frac{\partial}{\partial b_u} \left( r_{ui} - \mu - b_u - b_i \right)^2 = 2 \cdot \left( r_{ui} - \mu - b_u - b_i \right) \cdot \frac{\partial}{\partial b_u} \left( r_{ui} - \mu - b_u - b_i \right).
$$

#### Step 2: Differentiate $ g(x) $ with respect to $ b_u $
The expression $g(x) = r_{ui} - \mu - b_u - b_i $ is a sum of terms. We differentiate each term with respect to $b_u$:

- $r_{ui}$: It is a constant, its derivative is $0$.
- $ \mu $: It is a constant, its derivative is $ 0 $.
- $b_u$: Its derivative with respect to itself is $1$.
- $b_i$: It is a constant with respect to $b_u$, its derivative is $0$.

So:
$$
\frac{\partial}{\partial b_u} \left( r_{ui} - \mu - b_u - b_i \right) = -1.
$$

#### Step 3: Substitute into the chain rule
By substituting the derivative into the chain rule, we obtain:
$$
\frac{\partial}{\partial b_u} \left( r_{ui} - \mu - b_u - b_i \right)^2 = 2 \cdot \left( r_{ui} - \mu - b_u - b_i \right) \cdot (-1).
$$

The term $-1$ comes specifically from the derivative of $-b_u$ with respect to $b_u$.

This results in:
$$
-2 \sum_{(u, i) \in K} \left( r_{ui} - \mu - b_u - b_i \right).
$$

#### Regularization term:
The second term penalizes large values ​​of $b_u$. Its derivative is:
$$
\frac{\partial}{\partial b_u} \lambda \sum_u b_u^2 = 2 \lambda b_u.
$$

#### Combined formula:
By combining both terms, we obtain:
$$
\frac{\partial L}{\partial b_u} = -2 \sum_{(u, i) \in K} \left( r_{ui} - \mu - b_u - b_i \right) + 2 \lambda b_u.
$$

### 3. Simplified partial derivative with the error variable
If we define the error as:
$$
\text{error} = r_{ui} - \mu - b_u - b_i,
$$
The derivative simplifies to:
$$
\frac{\partial L}{\partial b_u} = -2 \cdot \text{error} + 2 \lambda b_u.
$$

### 4. Partial derivative with respect to $b_i$
The reasoning for \( b_i \) is symmetric. The partial derivative is:
$$
\frac{\partial L}{\partial b_i} = -2 \cdot \text{error} + 2 \lambda b_i.
$$

These derivatives guide the gradient descent to fit $b_u$ and $b_i$.


Optimizing $b_u $ and $ b_i $ with Gradient Descent
-------------
-----------

### Gradient Descent Steps
Now that we have the derivatives, we can perform gradient descent to optimize $ b_u $ and $ b_i $.

#### (a) Initialization:
- Initialize $ b_u $ and $ b_i $ to small random values or zeros.

#### (b) Update rules:
Using the gradient descent update formula:
$$
b_u \gets b_u - \eta \cdot \frac{\partial L}{\partial b_u},
$$
$$
b_i \gets b_i - \eta \cdot \frac{\partial L}{\partial b_i}.
$$
Here, $ \eta $ is the learning rate.

#### (c) Substitute the derivatives:
Substitute the partial derivatives:
$$
b_u \gets b_u + \eta \cdot \left( 2 \sum_{(u, i) \in K} \left( r_{ui} - \mu - b_u - b_i \right) - 2 \lambda b_u \right),
$$
$$
b_i \gets b_i + \eta \cdot \left( 2 \sum_{(u, i) \in K} \left( r_{ui} - \mu - b_u - b_i \right) - 2 \lambda b_i \right).
$$

#### (d) Iterate:
- Repeat the updates for a fixed number of iterations or until convergence (when changes in $ b_u $ and $ b_i $ are below a threshold).

---

### 5. Error interpretation
The term $ r_{ui} - \mu - b_u - b_i $ is the prediction error for a specific user $ u $ and item $ i $. Gradient descent minimizes this error while penalizing large values of $ b_u $ and $ b_i $ through the regularization term.

By iteratively applying these updates, the model adjusts $ b_u $ and $ b_i $ to better fit the data.


Practical Implementation
----------
---------

In [2]:
pip install scikit-surprise


     -------------------------------------- 154.4/154.4 kB 2.3 MB/s eta 0:00:00
  Installing build dependencies: started
  Installing build dependencies: finished with status 'done'
  Getting requirements to build wheel: started
  Getting requirements to build wheel: finished with status 'done'
  Preparing metadata (pyproject.toml): started
  Preparing metadata (pyproject.toml): finished with status 'done'
     -------------------------------------- 301.8/301.8 kB 9.4 MB/s eta 0:00:00
  Created wheel for scikit-surprise: filename=scikit_surprise-1.1.4-cp310-cp310-win_amd64.whl size=1296412 sha256=9e182d89f56eae762d11be4438b8aeb5d503a73501c680e7d264c743cd1a6db4
  Stored in directory: c:\users\brods\appdata\local\pip\cache\wheels\8a\ec\b2\58c47d2432188873013470d86ba25e2fc8d3d4e249ec625889
Successfully built scikit-surprise
  Attempting uninstall: joblib
    Found existing installation: joblib 1.1.1
    Uninstalling joblib-1.1.1:
      Successfully uninstalled joblib-1.1.1


This Python code implements a collaborative filtering model for movie recommendations using the MovieLens dataset.

The MovieLens 100K Dataset is a stable benchmark. It contains 100,000 ratings from 1000 users on 1700 movies and was released in 1998.

In this code, the model trains user and item biases through gradient descent and evaluates its performance based on RMSE (Root Mean Squared Error).

In [1]:
import numpy as np
from surprise import Dataset
from surprise.model_selection import train_test_split

# Load the MovieLens dataset provided by the Surprise library
data = Dataset.load_builtin('ml-100k')
trainset, testset = train_test_split(data, test_size=0.2) #the data is split into training and testing sets, using an 80/20 ratio

# Initialize parameters
mu = np.mean([r for (_, _, r) in trainset.all_ratings()])  # Global average rating
lambda_reg = 10  # Regularization parameter to prevent overfitting
learning_rate = 0.01  # Learning rate for gradient descent
num_epochs = 20  # Number of epochs for training

# Initialize user and item biases to zero
users = trainset.all_users()
items = trainset.all_items()
b_u = np.zeros(trainset.n_users)  # User biases
b_i = np.zeros(trainset.n_items)  # Item biases

# Training using gradient descent
for epoch in range(num_epochs):
    for u, i, r_ui in trainset.all_ratings():
        u, i = int(u), int(i)

        # Compute the error between the actual rating and the predicted rating
        pred = mu + b_u[u] + b_i[i]
        error = r_ui - pred

        # Update user and item biases using gradient descent
        b_u[u] += 2 * learning_rate * (error - lambda_reg * b_u[u])
        b_i[i] += 2 * learning_rate * (error - lambda_reg * b_i[i])

    # Calculate RMSE on the training set
    train_rmse = np.sqrt(np.mean([
        (r_ui - (mu + b_u[int(u)] + b_i[int(i)]))**2
        for u, i, r_ui in trainset.all_ratings()
    ]))
    print(f"Epoch {epoch + 1}/{num_epochs}, RMSE: {train_rmse:.4f}")

# Prediction function for the test set
def predict(user, item):
    """Predict the rating for a given user and item."""
    user_bias = b_u[user] if user < len(b_u) else 0
    item_bias = b_i[item] if item < len(b_i) else 0
    return mu + user_bias + item_bias

# Compute RMSE on the test set
test_rmse = np.sqrt(np.mean([
    (r_ui - predict(int(u), int(i)))**2
    for u, i, r_ui in testset
]))
print(f"Test RMSE: {test_rmse:.4f}")

Epoch 1/20, RMSE: 1.0909
Epoch 2/20, RMSE: 1.0904
Epoch 3/20, RMSE: 1.0903
Epoch 4/20, RMSE: 1.0902
Epoch 5/20, RMSE: 1.0901
Epoch 6/20, RMSE: 1.0901
Epoch 7/20, RMSE: 1.0901
Epoch 8/20, RMSE: 1.0901
Epoch 9/20, RMSE: 1.0901
Epoch 10/20, RMSE: 1.0901
Epoch 11/20, RMSE: 1.0900
Epoch 12/20, RMSE: 1.0900
Epoch 13/20, RMSE: 1.0900
Epoch 14/20, RMSE: 1.0900
Epoch 15/20, RMSE: 1.0900
Epoch 16/20, RMSE: 1.0900
Epoch 17/20, RMSE: 1.0900
Epoch 18/20, RMSE: 1.0900
Epoch 19/20, RMSE: 1.0900
Epoch 20/20, RMSE: 1.0900
Test RMSE: 1.1291


The results suggest that the model is not learning significantly after the first few iterations.

**Some points to keep in mind**

*1 - The RMSE on the training set*

The RMSE converges quickly and remains constant from the beginning. This could be due to:
- A high value of lambda regularization ($\lambda$) that restricts changes in $b_u$ and $b_i$ too much
- An insufficient step size ($\eta$), leading to slow and small updates to the parameters.
- A limited representation of the model, since you only adjust for biases without latent factors that can model more complex interactions between users and items (this is Koren's article main idea)

*2 -  The difference between the train and test set in the RMSE*

- The difference between the RMSE of training (∼1.0900) and test (∼1.1291) indicates possible overfitting to the training set.
- Although this is not severe, a more robust model with tight regularization could improve generalization.

# Hyperparameter optimization

In the first version of the code, we used only a combination of hyperparameters. In a real situation, choosing the most appropriate values for these elements requires careful research.

In this version of the code we implement a new strategy:
- A grid search is conducted over combinations of learning rates and regularization strengths
- For each combination, the model is trained, and the RMSE is calculated on both the training and test sets.
- The combination that minimizes test RMSE is recorded as the best configuration.
- Finally, it reports the best hyperparameters and the corresponding RMSE test.

In [3]:
import numpy as np
from surprise import Dataset
from surprise.model_selection import train_test_split
from itertools import product

# Load the MovieLens dataset
data = Dataset.load_builtin('ml-100k')
trainset, testset = train_test_split(data, test_size=0.2)

# Training and evaluation function
def train_and_evaluate(trainset, testset, learning_rate, lambda_reg, num_epochs=20):
    # Global mean of all ratings
    mu = np.mean([r for (_, _, r) in trainset.all_ratings()])

    # Initialize biases for users and items
    b_u = np.zeros(trainset.n_users)
    b_i = np.zeros(trainset.n_items)

    # Training loop with gradient descent
    for epoch in range(num_epochs):
        for u, i, r_ui in trainset.all_ratings():
            u, i = int(u), int(i)

            # Calculate prediction and error
            pred = mu + b_u[u] + b_i[i]
            error = r_ui - pred

            # Update user and item biases with gradient descent
            b_u[u] += 2 * learning_rate * (error - lambda_reg * b_u[u])
            b_i[i] += 2 * learning_rate * (error - lambda_reg * b_i[i])

            # Clip the values of b_u and b_i to avoid overflow
            b_u[u] = np.clip(b_u[u], -10, 10)
            b_i[i] = np.clip(b_i[i], -10, 10)

        # Calculate train RMSE for each epoch
        train_rmse = np.sqrt(np.mean([
            (r_ui - (mu + b_u[int(u)] + b_i[int(i)]))**2
            for u, i, r_ui in trainset.all_ratings()
        ]))
        print(f"Epoch {epoch + 1}/{num_epochs}, Train RMSE: {train_rmse:.4f}")

    # Prediction function for test data
    def predict(user, item):
        user_bias = b_u[user] if user < len(b_u) else 0
        item_bias = b_i[item] if item < len(b_i) else 0
        return mu + user_bias + item_bias

    # Calculate test RMSE
    test_rmse = np.sqrt(np.mean([
        (r_ui - predict(int(u), int(i)))**2
        for u, i, r_ui in testset
    ]))

    return train_rmse, test_rmse

# Grid search parameters
learning_rates = [0.005, 0.01, 0.05]
regularizations = [1, 10, 20]
num_epochs = 20

# Variables to store the best parameters
best_params = None
best_test_rmse = float('inf')

# Test all combinations of hyperparameters
for lr, reg in product(learning_rates, regularizations):
    print(f"Testing combination: learning_rate={lr}, lambda_reg={reg}")
    train_rmse, test_rmse = train_and_evaluate(trainset, testset, lr, reg, num_epochs)
    print(f"Train RMSE: {train_rmse:.4f}, Test RMSE: {test_rmse:.4f}")

    # Save the best parameters based on test RMSE
    if test_rmse < best_test_rmse:
        best_test_rmse = test_rmse
        best_params = (lr, reg)

print(f"\nBest hyperparameters: learning_rate={best_params[0]}, lambda_reg={best_params[1]}")
print(f"Test RMSE with best parameters: {best_test_rmse:.4f}")

Testing combination: learning_rate=0.005, lambda_reg=1
Epoch 1/20, Train RMSE: 1.0031
Epoch 2/20, Train RMSE: 0.9892
Epoch 3/20, Train RMSE: 0.9835
Epoch 4/20, Train RMSE: 0.9802
Epoch 5/20, Train RMSE: 0.9782
Epoch 6/20, Train RMSE: 0.9768
Epoch 7/20, Train RMSE: 0.9758
Epoch 8/20, Train RMSE: 0.9751
Epoch 9/20, Train RMSE: 0.9745
Epoch 10/20, Train RMSE: 0.9740
Epoch 11/20, Train RMSE: 0.9737
Epoch 12/20, Train RMSE: 0.9733
Epoch 13/20, Train RMSE: 0.9731
Epoch 14/20, Train RMSE: 0.9729
Epoch 15/20, Train RMSE: 0.9727
Epoch 16/20, Train RMSE: 0.9725
Epoch 17/20, Train RMSE: 0.9724
Epoch 18/20, Train RMSE: 0.9722
Epoch 19/20, Train RMSE: 0.9721
Epoch 20/20, Train RMSE: 0.9720
Train RMSE: 0.9720, Test RMSE: 1.1709
Testing combination: learning_rate=0.005, lambda_reg=10
Epoch 1/20, Train RMSE: 1.0927
Epoch 2/20, Train RMSE: 1.0918
Epoch 3/20, Train RMSE: 1.0915
Epoch 4/20, Train RMSE: 1.0913
Epoch 5/20, Train RMSE: 1.0912
Epoch 6/20, Train RMSE: 1.0911
Epoch 7/20, Train RMSE: 1.0911
Epo